pystac_client/collection_search.py

```diff
@@ -0,0 +1,521 @@
+                        "datetime, and q arguments are supported"
+                    )
+                    self._validate_client_side_args()
+                else:
+                    if not self._collection_search_free_text_enabled:
+                        warnings.warn(
+                            str(DoesNotConformTo("COLLECTION_SEARCH#FREE_TEXT"))
+                            + ". Free-text search is not enabled for collection search"
+                            "Free-text filters will be applied client-side."
+                        )
+
+        else:
+            self._stac_io = stac_io or StacApiIO()
+            self._collection_search_extension_enabled = (
+                collection_search_extension_enabled
+            )
+            self._collection_search_free_text_enabled = (
+                collection_search_free_text_enabled
+            )
+            self._validate_client_side_args()
+
+    def _validate_client_side_args(self) -> None:
+        """Client-side filtering only supports the bbox, datetime, and q parameters."""
+        args = {key for key, value in self._parameters.items() if value is not None}
+        extras = args - {"bbox", "datetime", "q", "limit", "max_items"}
+
+        if extras:
+            raise ValueError(
+                "Only the limit, max_collections, bbox, datetime, and q arguments are "
+                "supported for client-side filtering but these extra arguments were "
+                "provided: " + ",".join(extras)
+            )
+
+    @lru_cache(1)
+    def matched(self) -> Optional[int]:
+        """Return number matched for search
+
+        Returns the value from the `numberMatched` or `context.matched` field.
+        Not all APIs will support counts in which case a warning will be issued
+
+        Returns:
+            int: Total count of matched collections. If counts are not supported `None`
+            is returned.
+        """
+        found = None
+        iter = self.pages_as_dicts()
+        page = next(iter, None)
+
+        if not page:
+            return 0
+
+        # if collection search and free-text are fully supported, try reading a value
+        # from the search result context
+        if (
+            self._collection_search_extension_enabled
+            and self._collection_search_free_text_enabled
+        ):
+            if "context" in page:
+                found = page["context"].get("matched", None)
+            elif "numberMatched" in page:
+                found = page["numberMatched"]
+
+        if not found:
+            count = len(page["collections"])
+            for page in iter:
+                count += len(page["collections"])
+
+        # if context is missing or if collection search and free-text are not supported,
+        # just count the records
+        if found is None:
+            warnings.warn("numberMatched or context.matched not in response")
+            found = len(self.collection_list())
+
```

Comment:
```suggestion
```

I _think_ this is left over from my earlier suggestion?

Reply:
Good catch, I had combined two approaches in a bad way. Now I think it's right.